### PR TITLE
NEW multi select for customers categories filter in companies list

### PR DESCRIPTION
**Pull request by lvessiller-opendsi**
NEW multi select for customers categories filter in companies list
- change customer categories with a multi select filter
- filter with an "and" condition by default for selected customers categories
- add a checkbox to filter with an"or" condition for selected customers categories
- add "MAIN_SEARCH_CAT_OR_BY_DEFAULT" const to change default condition (OR by default instead of "AND")
- add  "MAIN_USE_WITH_INSTEAD_OF_INNER_JOIN" const to improve performances on "INNER JOIN" sql requests

### CODE REVIEW

File: htdocs/societe/list.php [hunk 1]
```diff
@@ -423,9 +449,33 @@
 $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."c_departements as state on (state.rowid = s.fk_departement)";
 $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."c_regions as region on (region.	code_region = state.fk_region)";
 // We'll need this table joined to the select in order to filter by categ
-if (!empty($search_categ_cus)) $sql .= ' LEFT JOIN '.MAIN_DB_PREFIX."categorie_societe as cc ON s.rowid = cc.fk_soc"; // We'll need this table joined to the select in order to filter by categ
+if (!empty($searchCategoryCustomerList)) {
+	if ($searchCategoryCustomerOperator == 1) {
```
Comment (eldy) [inline htdocs/societe/list.php:453]: To manage the AND, why not just add a criteria :
AND (select count(*) from categorie_societe where id in (A, B, C) where fk_soc =s.fk_soc) = 3
We will soon introduce an advanced filter selector to allow to combine any field with or and and and we must have simple rules to add such test that are compatible without changing the request.

Comment (lvessiller-opendsi) [reply htdocs/societe/list.php:454]: It's a good idea, I will try to replace this part of code

**Comment by lvessiller-opendsi** [comment]
@eldy 
conflicts resolved and add ADD criteria with count categories

File: htdocs/societe/list.php [hunk 2]
```diff
@@ -390,23 +404,59 @@
 	}
 } else dol_print_error($db);
 
-$sql = "SELECT s.rowid, s.nom as name, s.name_alias, s.barcode, s.address, s.town, s.zip, s.datec, s.code_client, s.code_fournisseur, s.logo,";
+$sql = "";
+if (!empty($conf->global->MAIN_USE_WITH_INSTEAD_OF_INNER_JOIN)) {
```
Comment (eldy) [inline htdocs/societe/list.php:408]: Such code will not be compatible with a future advanced filter editor.
Why not doing as suggested : No change on sql core request. We keep a simple sql request and the test is just added by added a where ?

Comment (lvessiller-opendsi) [reply htdocs/societe/list.php]: Ok I modified it

File: htdocs/societe/list.php [hunk 3]
```diff
@@ -390,23 +404,20 @@
 	}
 } else dol_print_error($db);
 
-$sql = "SELECT s.rowid, s.nom as name, s.name_alias, s.barcode, s.address, s.town, s.zip, s.datec, s.code_client, s.code_fournisseur, s.logo,";
+$sql  = "SELECT DISTINCT s.rowid, s.nom as name, s.name_alias, s.barcode, s.address, s.town, s.zip, s.datec, s.code_client, s.code_fournisseur, s.logo,";
```
Comment (eldy) [inline htdocs/societe/list.php:407]: The distinct should not be required.
If we search on 1 category, we can have a left join on the id of category, so we will have no duplicated record.
And if we need a AND, the left join is no more required, only a AND added into where will be enough (a subrequest to test a condition on id of company into categorie_societe)

Comment (lvessiller-opendsi) [reply htdocs/societe/list.php]: You're right, it's no longer appears to be necessary.

Comment (lvessiller-opendsi) [reply htdocs/societe/list.php:407]: Sorry, it's useful for OR operator in categories filters

**Comment by lvessiller-opendsi** [comment]
@eldy 
Conflicts resolved

**Comment by lvessiller-opendsi** [comment]
I close this PR because it's too old and I will do again with the new way to have customer categories filter in lists.

**Pull request closed without merging**